sdks/java/io/iceberg/src/test/java/org/apache/beam/sdk/io/iceberg/IcebergIOIT.java

```diff
@@ -0,0 +1,266 @@
+import org.apache.iceberg.hadoop.HadoopCatalog;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.DataWriter;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.parquet.Parquet;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class IcebergIOIT implements Serializable {
+
+  public interface IcebergIOTestPipelineOptions extends GcpOptions {
+    @Description("Number of records that will be written and/or read by the test")
+    @Default.Integer(1000) // deliberately small so no-args execution is quick
+    Integer getNumRecords();
+
+    void setNumRecords(Integer numRecords);
+
+    @Description("Number of shards in the test table")
+    @Default.Integer(10)
+    Integer getNumShards();
+
+    void setNumShards(Integer numShards);
+  }
+
+  @Rule public TestPipeline writePipeline = TestPipeline.create();
+
+  @Rule public TestPipeline readPipeline = TestPipeline.create();
+
+  static IcebergIOTestPipelineOptions options;
+
+  static Configuration catalogHadoopConf;
+
+  @BeforeClass
+  public static void beforeClass() {
+    PipelineOptionsFactory.register(IcebergIOTestPipelineOptions.class);
+    options = TestPipeline.testingPipelineOptions().as(IcebergIOTestPipelineOptions.class);
+
+    catalogHadoopConf = new Configuration();
+    catalogHadoopConf.set("fs.gs.project.id", options.getProject());
+    catalogHadoopConf.set("fs.gs.auth.type", "SERVICE_ACCOUNT_JSON_KEYFILE");
+    catalogHadoopConf.set(
+        "fs.gs.auth.service.account.json.keyfile", System.getenv("GOOGLE_APPLICATION_CREDENTIALS"));
+  }
+
+  static final org.apache.beam.sdk.schemas.Schema BEAM_SCHEMA =
+      org.apache.beam.sdk.schemas.Schema.builder()
+          .addInt32Field("int")
+          .addFloatField("float")
+          .addDoubleField("double")
+          .addInt64Field("long")
+          .addStringField("str")
+          .addBooleanField("bool")
+          .addByteArrayField("bytes")
+          .build();
+
+  static final Schema ICEBERG_SCHEMA =
+      SchemaAndRowConversions.beamSchemaToIcebergSchema(BEAM_SCHEMA);
+
+  Map<String, Object> getValues(int num) {
+    String strNum = Integer.toString(num);
+    return ImmutableMap.<String, Object>builder()
+        .put("int", num)
+        .put("float", Float.valueOf(strNum))
+        .put("double", Double.valueOf(strNum))
+        .put("long", Long.valueOf(strNum))
+        .put("str", strNum)
+        .put("bool", num % 2 == 0)
+        .put("bytes", ByteBuffer.wrap(new byte[] {(byte) num}))
+        .build();
+  }
+
+  /**
+   * Populates the Iceberg table according to the configuration specified in {@link
+   * IcebergIOTestPipelineOptions}. Returns a {@link List<Row>} of expected elements.
+   */
+  List<Row> populateTable(Table table) throws IOException {
```

Comment:
Problem here is that this method is not scale-independent. You can't run the `integrationTest` job on e.g. Dataflow with pipeline options that scale up too big for direct runner. What you have done is definitely useful as-is. But what we also really need are tests where we don't rely on being able to hold the expected results in memory. So you need to calculate a signature/checksum of some sort and test against that when you move to a scale-independent test.